In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import csv

## 1. Look through the data

1.1 Company name

In [2]:
##### Read company names into a dictionary
def readNamesIntoDict():
    d = dict()
    input_file = csv.DictReader(open("SP_500_firms.csv"))
    for row in input_file:
        #print(row)
        d[row['Symbol']] = [row['Name'],row['Sector']]
    return d

In [43]:
namesDict = readNamesIntoDict()
namesDict

{'A': ['Agilent Technologies Inc', 'Health Care'],
 'AA': ['Alcoa Inc', 'Materials'],
 'AAL': ['American Airlines Group', 'Industrials'],
 'AAP': ['Advance Auto Parts', 'Consumer Discretionary'],
 'AAPL': ['Apple Inc.', 'Information Technology'],
 'ABBV': ['AbbVie', 'Health Care'],
 'ABC': ['AmerisourceBergen Corp', 'Health Care'],
 'ABT': ['Abbott Laboratories', 'Health Care'],
 'ACN': ['Accenture plc', 'Information Technology'],
 'ADBE': ['Adobe Systems Inc', 'Information Technology'],
 'ADI': ['Analog Devices, Inc.', 'Information Technology'],
 'ADM': ['Archer-Daniels-Midland Co', 'Consumer Staples'],
 'ADP': ['Automatic Data Processing', 'Information Technology'],
 'ADS': ['Alliance Data Systems', 'Information Technology'],
 'ADSK': ['Autodesk Inc', 'Information Technology'],
 'AEE': ['Ameren Corp', 'Utilities'],
 'AEP': ['American Electric Power', 'Utilities'],
 'AES': ['AES Corp', 'Utilities'],
 'AET': ['Aetna Inc', 'Health Care'],
 'AFL': ['AFLAC Inc', 'Financials'],
 'AGN': ['A

In [4]:
compNames = namesDict.keys()
compNames

dict_keys(['BAC', 'CCL', 'RSG', 'SBUX', 'SLG', 'WFC', 'HRB', 'AMAT', 'SYF', 'PM', 'JNJ', 'BWA', 'RRC', 'ETR', 'HPQ', 'DTE', 'TWX', 'MU', 'MLM', 'LUV', 'FTI', 'CMS', 'EMR', 'BEN', 'COG', 'KMI', 'MMM', 'UNH', 'EMN', 'TSCO', 'NRG', 'D', 'CMI', 'CSX', 'T', 'PSA', 'CI', 'TMK', 'BAX', 'ABBV', 'TIF', 'PLD', 'PCAR', 'VMC', 'RF', 'QCOM', 'FAST', 'RTN', 'HRL', 'VNO', 'WLTW', 'AEP', 'EQR', 'LEG', 'CMG', 'DFS', 'HP', 'GOOG', 'LEN', 'PBCT', 'STI', 'TXT', 'CHRW', 'DLPH', 'GPC', 'MRO', 'SWK', 'AMG', 'DIS', 'CSCO', 'ABC', 'AFL', 'MET', 'AA', 'MCK', 'PYPL', 'DVN', 'FOXA', 'TSO', 'COF', 'DOW', 'SEE', 'EIX', 'FRT', 'UHS', 'VIAB', 'NSC', 'NOV', 'SWKS', 'PCLN', 'TGNA', 'GWW', 'STX', 'CXO', 'WU', 'PEG', 'HCP', 'TDC', 'PSX', 'URI', 'NEM', 'AIZ', 'EA', 'HSY', 'CBS', 'AKAM', 'EXR', 'AEE', 'OXY', 'V', 'STT', 'SIG', 'SO', 'HD', 'TSS', 'FLR', 'PNR', 'ALB', 'CLX', 'KSU', 'GIS', 'AMP', 'FB', 'EMC', 'WHR', 'AES', 'SYK', 'ATVI', 'PRGO', 'IPG', 'C', 'XOM', 'XLNX', 'SLB', 'LM', 'EW', 'DHR', 'YUM', 'VZ', 'XYL', 'SNA', '

In [5]:
print(len(namesDict)) # we have 504 companys

504


1.2 Company stock prices

In [6]:
##### Prices into standarad Python data structures

# Read prices into dictionary of lists

def readPricesIntoDict():
    input_file = csv.DictReader(open('SP_500_close_2015.csv', 'r')) 
    d = dict()
    for row in input_file:
        for column, value in row.items():
            d.setdefault(column, []).append(value)
    return d


prices = readPricesIntoDict()

In [45]:
prices['A']

['39.911784999999995',
 '39.163928999999996',
 '38.553838',
 '39.065528',
 '40.236509000000005',
 '39.941303999999995',
 '39.468976',
 '38.917924',
 '38.435757',
 '37.402535',
 '37.638701',
 '37.323815',
 '37.550139',
 '39.016328',
 '38.189753',
 '38.524319',
 '38.13071',
 '37.392696',
 '37.845344',
 '37.166373',
 '38.071668',
 '38.986805',
 '38.484957',
 '38.898243',
 '38.711281',
 '38.416076000000004',
 '39.036004999999996',
 '38.701443',
 '39.380414',
 '39.508337',
 '39.872423',
 '40.876121999999995',
 '40.492356',
 '41.27957',
 '41.063084',
 '41.387812',
 '41.525574',
 '41.683017',
 '41.535413',
 '42.017583',
 '41.584613',
 '41.32877',
 '41.545255',
 '40.866279999999996',
 '41.072927',
 '39.980666',
 '40.197146999999994',
 '40.443152000000005',
 '40.216828',
 '41.141808000000005',
 '40.915484',
 '41.446851',
 '41.535413',
 '41.535413',
 '41.525574',
 '40.433313',
 '40.157789',
 '40.049547',
 '40.55263',
 '41.15436',
 '40.986663',
 '40.828833',
 '41.479884000000006',
 '41.361512',
 

In [8]:
len(prices)

497

In [9]:
##### Prices into pandas

# Open data with pandas 
filename = 'SP_500_close_2015.csv'
priceData = pd.read_csv(filename,index_col = 0)

print(type(priceData))
print(priceData.columns)


<class 'pandas.core.frame.DataFrame'>
Index(['MMM', 'ABT', 'ABBV', 'ACN', 'ATVI', 'AYI', 'ADBE', 'AAP', 'AES', 'AET',
       ...
       'XEL', 'XRX', 'XLNX', 'XL', 'XYL', 'YHOO', 'YUM', 'ZBH', 'ZION', 'ZTS'],
      dtype='object', length=496)


In [10]:
str(priceData)

'                   MMM        ABT       ABBV         ACN       ATVI  \\\nDate                                                                  \n2015-01-02  156.678596  43.160459  61.986410   86.129228  19.765196   \n2015-01-05  153.145069  43.170070  60.819874   84.674997  19.490271   \n2015-01-06  151.511999  42.679830  60.518833   84.064223  19.126976   \n2015-01-07  152.610267  43.025880  62.964797   85.828689  18.714587   \n2015-01-08  156.267949  43.910238  63.623323   87.137495  18.901144   \n2015-01-09  154.348374  43.448835  61.882926   87.040548  18.557487   \n2015-01-12  153.507977  43.814114  61.864114   86.778790  18.518213   \n2015-01-13  153.393366  43.108681  60.082211   87.089025  18.223648   \n2015-01-14  152.648461  42.789785  60.376037   86.071061  18.586944   \n2015-01-15  152.476567  42.470893  59.608302   85.479673  18.213829   \n2015-01-16  154.711282  42.992720  61.172204   86.177705  19.883022   \n2015-01-20  154.653983  42.731807  60.139082   85.915947  19.8

In [11]:
priceData.head()

MMM        ABT       ABBV        ACN       ATVI  \
Date                                                                 
2015-01-02  156.678596  43.160459  61.986410  86.129228  19.765196   
2015-01-05  153.145069  43.170070  60.819874  84.674997  19.490271   
2015-01-06  151.511999  42.679830  60.518833  84.064223  19.126976   
2015-01-07  152.610267  43.025880  62.964797  85.828689  18.714587   
2015-01-08  156.267949  43.910238  63.623323  87.137495  18.901144   

                   AYI       ADBE         AAP        AES        AET  \
Date                                                                  
2015-01-02  139.234407  72.339996  158.132353  12.860543  87.354435   
2015-01-05  135.889914  71.980003  156.047994  12.494440  86.173965   
2015-01-06  134.187800  70.529999  155.938290  12.212822  86.301853   
2015-01-07  136.566769  71.110001  159.289228  12.231597  88.033197   
2015-01-08  141.344618  72.919998  160.685446  12.419342  90.885990   

              ...            XEL        XRX       XLNX         XL        XYL  \
Date          ...                                                              
2015-01-02    ...      33.927597  13.191375  41.513829  32.998169  37.072501   
2015-01-05    ...      33.542483  12.903563  40.752108  33.199318  34.765202   
2015-01-06    ...      33.720953  12.730876  40.161775  33.362155  34.560759   
2015-01-07    ...      34.012134  12.893969  40.180818  33.381309  34.833350   
2015-01-08    ...      34.406643  13.248937  41.075840  33.927287  35.096207   

                 YHOO        YUM         ZBH       ZION        ZTS  
Date                                                                
2015-01-02  50.169998  69.542301  110.963875  27.853656  42.723658  
2015-01-05  49.130001  68.129346  115.103219  26.810004  42.467175  
2015-01-06  49.209999  67.293113  114.127520  25.786046  42.052863  
2015-01-07  48.590000  69.523080  116.965922  26.032190  42.920947  
2015-01-08  50.230000  70.734178  118.207726  26.396482  43.581878  

[5 rows x 496 columns]

In [12]:
priceData.describe()

MMM         ABT        ABBV         ACN        ATVI         AYI  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean   151.240445   45.067555   59.532853   95.540953   27.351820  185.839553   
std      7.107680    2.522206    4.742366    6.875068    5.686488   24.958500   
min    133.969117   38.121817   46.901920   81.465997   18.213829  134.187800   
25%    146.098763   43.318241   55.582699   90.824246   22.882658  166.476869   
50%    152.868585   44.921188   59.153699   94.982332   25.408236  182.209625   
75%    156.287518   47.124548   63.470802  101.395491   31.392041  201.155203   
max    163.841622   49.970225   68.578266  108.674818   39.260511  239.748197   

             ADBE         AAP         AES         AET     ...             XEL  \
count  252.000000  252.000000  252.000000  252.000000     ...      252.000000   
mean    80.967817  163.313504   11.452325  107.930044     ...       33.192459   
std      6.373760   14.847493    1.288663    8.775297     ...        1.329804   
min     69.989998  142.671386    8.562809   86.173965     ...       30.564587   
25%     76.490002  151.131895   10.455896  104.172749     ...       32.196438   
50%     79.825001  159.705316   11.926511  108.143008     ...       33.028862   
75%     84.069998  173.011927   12.531814  114.007040     ...       34.326218   
max     95.559998  200.062855   13.292200  131.110696     ...       35.909463   

              XRX        XLNX          XL         XYL        YHOO         YUM  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean    11.069191   42.454640   36.025538   34.701831   38.971944   78.129855   
std      1.234858    3.035655    1.275883    1.549875    5.631780    7.133025   
min      9.081719   36.734029   32.998169   29.885909   27.600000   65.847064   
25%     10.100338   40.161342   35.395930   33.866610   33.460000   71.694150   
50%     10.664143   41.758568   36.091841   34.888744   39.355002   77.018419   
75%     12.388123   45.348611   36.777727   35.894561   43.884999   85.664143   
max     13.431218   49.220191   39.467252   37.264386   50.230000   92.198695   

              ZBH        ZION         ZTS  
count  252.000000  252.000000  252.000000  
mean   107.543330   28.180492   45.665167  
std      7.499312    1.982506    2.258266  
min     91.636308   23.501830   39.324541  
25%    101.314589   26.931264   43.978874  
50%    107.835618   27.954411   45.770399  
75%    114.208823   29.715322   47.099767  
max    120.001439   32.103599   54.831212  

[8 rows x 496 columns]

In [46]:
firstPrices =priceData.ix[0] # This is a "series" of first-day prices
print(type(firstPrices))
firstColumnPrices = priceData.ix[:,0] # First company by index
firstColumnPrices.head()

<class 'pandas.core.series.Series'>


Date
2015-01-02    156.678596
2015-01-05    153.145069
2015-01-06    151.511999
2015-01-07    152.610267
2015-01-08    156.267949
Name: MMM, dtype: float64

In [47]:
applePrices = priceData['AAPL'] # Get by column name
msftPrices = priceData['MSFT']
#print(applePrices)
#print(msftPrices)

In [49]:
# Create dataframe from series, then add another series
customPrices = applePrices.to_frame('AAPL')
customPrices = customPrices.join(msftPrices.to_frame('MSFT'))

customPrices.head()

AAPL       MSFT
Date                             
2015-01-02  105.698620  44.604322
2015-01-05  102.720920  44.194149
2015-01-06  102.730590  43.545497
2015-01-07  104.171098  44.098757
2015-01-08  108.173588  45.396060

In [16]:
# Normalise data by first price
pricesScaled = priceData.divide(priceData.ix[0]) 
pricesScaled.head()

MMM       ABT      ABBV       ACN      ATVI       AYI  \
Date                                                                     
2015-01-02  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000   
2015-01-05  0.977447  1.000223  0.981181  0.983116  0.986090  0.975979   
2015-01-06  0.967024  0.988864  0.976324  0.976024  0.967710  0.963755   
2015-01-07  0.974034  0.996882  1.015784  0.996511  0.946846  0.980841   
2015-01-08  0.997379  1.017372  1.026408  1.011706  0.956284  1.015156   

                ADBE       AAP       AES       AET    ...          XEL  \
Date                                                  ...                
2015-01-02  1.000000  1.000000  1.000000  1.000000    ...     1.000000   
2015-01-05  0.995024  0.986819  0.971533  0.986486    ...     0.988649   
2015-01-06  0.974979  0.986125  0.949635  0.987950    ...     0.993909   
2015-01-07  0.982997  1.007316  0.951095  1.007770    ...     1.002492   
2015-01-08  1.008018  1.016145  0.965693  1.040428    ...     1.014120   

                 XRX      XLNX        XL       XYL      YHOO       YUM  \
Date                                                                     
2015-01-02  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000   
2015-01-05  0.978182  0.981651  1.006096  0.937763  0.979271  0.979682   
2015-01-06  0.965091  0.967431  1.011030  0.932248  0.980865  0.967657   
2015-01-07  0.977455  0.967890  1.011611  0.939601  0.968507  0.999724   
2015-01-08  1.004364  0.989450  1.028157  0.946691  1.001196  1.017139   

                 ZBH      ZION       ZTS  
Date                                      
2015-01-02  1.000000  1.000000  1.000000  
2015-01-05  1.037304  0.962531  0.993997  
2015-01-06  1.028511  0.925769  0.984299  
2015-01-07  1.054090  0.934606  1.004618  
2015-01-08  1.065281  0.947685  1.020088  

[5 rows x 496 columns]

In [17]:
# Plot
priceFig = pricesScaled.plot(legend=False,figsize=(6,4))
priceFig
# Save figure into working directory
#plt.savefig('stocks2015.png', bbox_inches='tight')

In [18]:
# Loop through companies
for index,company in enumerate(priceData.columns):
        print(company,index)

MMM 0
ABT 1
ABBV 2
ACN 3
ATVI 4
AYI 5
ADBE 6
AAP 7
AES 8
AET 9
AMG 10
AFL 11
A 12
APD 13
AKAM 14
ALK 15
ALB 16
AA 17
ALXN 18
ALLE 19
AGN 20
ADS 21
LNT 22
ALL 23
GOOGL 24
GOOG 25
MO 26
AMZN 27
AEE 28
AAL 29
AEP 30
AXP 31
AIG 32
AMT 33
AWK 34
AMP 35
ABC 36
AME 37
AMGN 38
APH 39
APC 40
ADI 41
ANTM 42
AON 43
APA 44
AIV 45
AAPL 46
AMAT 47
ADM 48
AJG 49
AIZ 50
T 51
ADSK 52
ADP 53
AN 54
AZO 55
AVGO 56
AVB 57
AVY 58
BHI 59
BLL 60
BAC 61
BCR 62
BAX 63
BBT 64
BDX 65
BBBY 66
BRK-B 67
BBY 68
BIIB 69
BLK 70
HRB 71
BA 72
BWA 73
BXP 74
BSX 75
BMY 76
BF-B 77
CHRW 78
CA 79
COG 80
CPB 81
COF 82
CAH 83
KMX 84
CCL 85
CAT 86
CBG 87
CBS 88
CELG 89
CNC 90
CNP 91
CTL 92
CERN 93
CF 94
SCHW 95
CHK 96
CVX 97
CMG 98
CB 99
CHD 100
CI 101
XEC 102
CINF 103
CTAS 104
CSCO 105
C 106
CFG 107
CTXS 108
CME 109
CMS 110
COH 111
CTSH 112
CL 113
CMCSA 114
CMA 115
CAG 116
CXO 117
COP 118
ED 119
STZ 120
GLW 121
COST 122
CCI 123
CSX 124
CMI 125
CVS 126
DHI 127
DHR 128
DRI 129
DVA 130
DE 131
DLPH 132
DAL 133
XRAY 134
DVN 135
DO 1

In [19]:
len(priceData.columns)

496

## 2. Stock Analysis
### 2.1 Analyse which stocks perform similarly over time
Our features are the daily stock prices, The year return is just not so procise. We need to analyse **daily stock price movements** and get the **daily return of the stock**

In [20]:
priceDailyReturn = pd.DataFrame.copy(priceData) # can't directly assign the priceData to priceDailyReturn dataframe
priceDailyReturn.iloc[:,:] = 0  # make the dataframe elements all to 0
# for loop for every row, shape[0] means row numbers, shape[1] means colums numbers
for i in range(1,int(priceDailyReturn.shape[0])):
    priceDailyReturn.iloc[i,:] = (priceData.iloc[i,:] - priceData.iloc[i-1,:]) / priceData.iloc[i-1,:]     
priceDailyReturn.to_csv("priveDailyReturn.csv")
priceDailyReturn.head()

MMM       ABT      ABBV       ACN      ATVI       AYI  \
Date                                                                     
2015-01-02  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2015-01-05 -0.022553  0.000223 -0.018819 -0.016884 -0.013910 -0.024021   
2015-01-06 -0.010664 -0.011356 -0.004950 -0.007213 -0.018640 -0.012526   
2015-01-07  0.007249  0.008108  0.040417  0.020989 -0.021561  0.017729   
2015-01-08  0.023967  0.020554  0.010459  0.015249  0.009969  0.034985   

                ADBE       AAP       AES       AET    ...          XEL  \
Date                                                  ...                
2015-01-02  0.000000  0.000000  0.000000  0.000000    ...     0.000000   
2015-01-05 -0.004976 -0.013181 -0.028467 -0.013514    ...    -0.011351   
2015-01-06 -0.020145 -0.000703 -0.022539  0.001484    ...     0.005321   
2015-01-07  0.008223  0.021489  0.001537  0.020061    ...     0.008635   
2015-01-08  0.025453  0.008765  0.015349  0.032406    ...     0.011599   

                 XRX      XLNX        XL       XYL      YHOO       YUM  \
Date                                                                     
2015-01-02  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2015-01-05 -0.021818 -0.018349  0.006096 -0.062237 -0.020729 -0.020318   
2015-01-06 -0.013383 -0.014486  0.004905 -0.005881  0.001628 -0.012274   
2015-01-07  0.012811  0.000474  0.000574  0.007887 -0.012599  0.033138   
2015-01-08  0.027530  0.022275  0.016356  0.007546  0.033752  0.017420   

                 ZBH      ZION       ZTS  
Date                                      
2015-01-02  0.000000  0.000000  0.000000  
2015-01-05  0.037304 -0.037469 -0.006003  
2015-01-06 -0.008477 -0.038193 -0.009756  
2015-01-07  0.024870  0.009546  0.020643  
2015-01-08  0.010617  0.013994  0.015399  

[5 rows x 496 columns]

First, we describe the daily returns for all stocks in the data

In [21]:
pdr_describe = priceDailyReturn.describe()
pdr_describe

MMM         ABT        ABBV         ACN        ATVI         AYI  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean    -0.000168    0.000178   -0.000093    0.000811    0.002789    0.002173   
std      0.011636    0.013721    0.020014    0.012870    0.017507    0.015613   
min     -0.060268   -0.072777   -0.103288   -0.053355   -0.063169   -0.048394   
25%     -0.006604   -0.006618   -0.011308   -0.006258   -0.006138   -0.007628   
50%      0.000234    0.000099    0.001177    0.000618    0.002506    0.001506   
75%      0.006016    0.008361    0.010136    0.007365    0.011021    0.011010   
max      0.041249    0.042617    0.100739    0.067566    0.118426    0.071632   

             ADBE         AAP         AES         AET     ...             XEL  \
count  252.000000  252.000000  252.000000  252.000000     ...      252.000000   
mean     0.001146   -0.000009   -0.001134    0.000987     ...        0.000198   
std      0.014766    0.019470    0.017846    0.018285     ...        0.011881   
min     -0.058324   -0.154261   -0.086876   -0.065226     ...       -0.049255   
25%     -0.007258   -0.006503   -0.011945   -0.008872     ...       -0.006656   
50%      0.002456    0.000663   -0.000447    0.001928     ...        0.001435   
75%      0.009962    0.008110    0.009801    0.011526     ...        0.008593   
max      0.037586    0.111417    0.057490    0.046229     ...        0.031543   

              XRX        XLNX          XL         XYL        YHOO         YUM  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean    -0.000758    0.000526    0.000639   -0.000019   -0.001441    0.000342   
std      0.017959    0.015408    0.010174    0.013108    0.019437    0.020380   
min     -0.087519   -0.061431   -0.035901   -0.062237   -0.076195   -0.188324   
25%     -0.010464   -0.008938   -0.005863   -0.007078   -0.011532   -0.008299   
50%     -0.000378    0.000559    0.000965   -0.001103   -0.001913   -0.000437   
75%      0.009967    0.009210    0.006438    0.008422    0.010028    0.009643   
max      0.065489    0.067132    0.031662    0.034612    0.062262    0.069102   

              ZBH        ZION         ZTS  
count  252.000000  252.000000  252.000000  
mean    -0.000254    0.000041    0.000603  
std      0.012862    0.017464    0.018589  
min     -0.047685   -0.046207   -0.122066  
25%     -0.007809   -0.009895   -0.008379  
50%     -0.000135    0.000000    0.000100  
75%      0.007122    0.011570    0.008861  
max      0.071523    0.067133    0.114062  

[8 rows x 496 columns]

Then we find the company with the minimum and maximum daily return 

In [22]:
#find the maximum daily return
print('The maximum daily return company is ' + pdr_describe.loc['max'].idxmax())
print(pdr_describe.loc['max'].max())
#find the minimum daily return
print('The minimum daily return company is ' + pdr_describe.loc['min'].idxmin())
print(pdr_describe.loc['min'].min())

The maximum daily return company is FCX
0.286616201466
The minimum daily return company is PWR
-0.285005695727


In [23]:
#find the overall best company
print('The overall best company is ' +pdr_describe.loc['mean'].idxmax())
print(pdr_describe.loc['mean'].max())
#find the overall worst company
print('The overall worst company is ' +pdr_describe.loc['mean'].idxmin())
print(pdr_describe.loc['mean'].min())

The overall best company is NFLX
0.00378946264328
The overall worst company is CHK
-0.0049070365421


In [24]:
#find the most volatility company
print('The most volatility company is ' + pdr_describe.loc['std'].idxmax())
print(pdr_describe.loc['std'].max())
#find the least volatility company
print('The least volatility company is ' + pdr_describe.loc['std'].idxmin())
print(pdr_describe.loc['std'].min())

The most volatility company is FCX
0.0439840636135
The least volatility company is KO
0.00904486717773


## 2. Correlation
The calculation of each correlation was completed in the previous part. Using that data, we started with assinging the firm names and correlation into a pandas dataframe.

In [25]:
filename2 = 'SP_500_firms.csv'
firms = pd.read_csv(filename2)
firms.head()
#priceDailyReturnName = priceDailyReturn.coclumns
#priceDailyReturn.columns = list(firms['Name'])
#priceDailyReturn.columns

Symbol                 Name                  Sector
0    MMM           3M Company             Industrials
1    ABT  Abbott Laboratories             Health Care
2   ABBV               AbbVie             Health Care
3    ACN        Accenture plc  Information Technology
4   ATVI  Activision Blizzard  Information Technology

In [26]:
cor_price = priceDailyReturn.corr() # for pairwise correlation of dataframe columns
cor_price.head(7)

MMM       ABT      ABBV       ACN      ATVI       AYI      ADBE  \
MMM   1.000000  0.598665  0.322637  0.632050  0.418519  0.450872  0.468745   
ABT   0.598665  1.000000  0.483668  0.644078  0.434357  0.494303  0.585668   
ABBV  0.322637  0.483668  1.000000  0.318048  0.226854  0.269083  0.347331   
ACN   0.632050  0.644078  0.318048  1.000000  0.462558  0.481309  0.560768   
ATVI  0.418519  0.434357  0.226854  0.462558  1.000000  0.310525  0.482797   
AYI   0.450872  0.494303  0.269083  0.481309  0.310525  1.000000  0.341863   
ADBE  0.468745  0.585668  0.347331  0.560768  0.482797  0.341863  1.000000   

           AAP       AES       AET    ...          XEL       XRX      XLNX  \
MMM   0.257132  0.335379  0.317369    ...     0.409629  0.499817  0.507627   
ABT   0.371080  0.450206  0.464380    ...     0.453393  0.538238  0.507487   
ABBV  0.292497  0.303988  0.419361    ...     0.261255  0.291590  0.323336   
ACN   0.303224  0.395808  0.401344    ...     0.388994  0.531869  0.510195   
ATVI  0.223472  0.198081  0.285139    ...     0.295286  0.369710  0.398587   
AYI   0.281292  0.311834  0.322694    ...     0.292764  0.443355  0.403896   
ADBE  0.291740  0.283060  0.403982    ...     0.339726  0.461567  0.460718   

            XL       XYL      YHOO       YUM       ZBH      ZION       ZTS  
MMM   0.543725  0.623740  0.316707  0.356203  0.433634  0.486263  0.293895  
ABT   0.619808  0.567308  0.376218  0.388071  0.640906  0.512479  0.463926  
ABBV  0.267432  0.321926  0.307018  0.289727  0.508626  0.323670  0.406245  
ACN   0.563160  0.615344  0.431342  0.353530  0.461179  0.546650  0.414054  
ATVI  0.388844  0.396983  0.365830  0.330461  0.370655  0.346774  0.323145  
AYI   0.478268  0.456437  0.330368  0.231670  0.373889  0.292102  0.306795  
ADBE  0.492877  0.476534  0.313914  0.482703  0.409579  0.445168  0.377819  

[7 rows x 496 columns]

Here we build our own correlation function to check the result above

In [27]:
def callCor(vec1, vec2):
    """
    input: two vector array
    output: the correlation scores
    """
    vec_idx = range(len(vec1)) # the length of vector
    
    vec1_sum = sum([vec1[i] for i in vec_idx]) # the sum of vector1
    vec2_sum = sum([vec2[i] for i in vec_idx]) # the sum of vector2
    
    vec1_square_sum = sum([pow(vec1[i], 2) for i in vec_idx]) # the sum of vector1 square
    vec2_square_sum = sum([pow(vec2[i], 2) for i in vec_idx]) # the sum of vector2 square
    
    product_sum = sum([vec1[i] * vec2[i] for i in vec_idx]) # the sum of product of vector1 and vector 2
    
    numerator = len(vec1) * product_sum - vec1_sum * vec2_sum
    denominator = (len(vec1) * vec1_square_sum - vec1_sum ** 2)**0.5 * (len(vec1) * vec2_square_sum - vec2_sum ** 2)**0.5
    
    if denominator == 0:
        return 0
    
    result = numerator / denominator
    
    return result

callCor(priceDailyReturn.iloc[:,0],priceDailyReturn.iloc[:,1])

0.59866496204887343

In [29]:
def dfCor(df):
    """
    input: dataframe
    output: the pairwise correlation of dataframe
    """
    colname = list(df.columns) # the column name of dataframe
    
    data = pd.DataFrame(np.zeros([len(colname), len(colname)]), columns= colname, index= colname)
    
    for i in colname:
        for j in colname:
            data.loc[i,j] = callCor(df[i],df[j])
    
    return data

dfCor(priceDailyReturn)
            

KeyboardInterrupt: 

After getting all the tables, we created functions to call the full name of a company using its ticker only. There are two ways that we explored in this part. Firstly, the names and the correlation were taken from 2 different tables using a `callName` function and `callCor` function. Secondly, the two tables loaded in the console were merged together to get one table with all the information needed. Then, `callName2` function was created to smoothen the merging. Then, `callCor2` was defined to get the correlation and both names of the companies. 

In [30]:
def callName(n):
    x = firms[firms.Symbol == str(n)].Name.item()
    return str(x)
callName('ABT')

'Abbott Laboratories'

In [31]:
def callName2(n):
    x = namesDict[n][0]
    return x
callName2('ABT')  

'Abbott Laboratories'

In [32]:
def callCor(a, b):
    cor = cor_price[a][b]
    print('The correlation between %s and %s is %s' % (callName(a), callName(b), cor))

callCor('ABT', 'AYI')

The correlation between Abbott Laboratories and Acuity Brands Inc is 0.494302519137


In [33]:
# Merge the name of each company into price dataframe
cor_price_name = cor_price.copy()
cor_price_name['Name'] = ""

for i in list(cor_price_name):
    if i == 'Name':
        pass
    else:
        cor_price_name.set_value(i, 'Name', callName2(i))

cor_price_name.head()

MMM       ABT      ABBV       ACN      ATVI       AYI      ADBE  \
MMM   1.000000  0.598665  0.322637  0.632050  0.418519  0.450872  0.468745   
ABT   0.598665  1.000000  0.483668  0.644078  0.434357  0.494303  0.585668   
ABBV  0.322637  0.483668  1.000000  0.318048  0.226854  0.269083  0.347331   
ACN   0.632050  0.644078  0.318048  1.000000  0.462558  0.481309  0.560768   
ATVI  0.418519  0.434357  0.226854  0.462558  1.000000  0.310525  0.482797   

           AAP       AES       AET         ...                XRX      XLNX  \
MMM   0.257132  0.335379  0.317369         ...           0.499817  0.507627   
ABT   0.371080  0.450206  0.464380         ...           0.538238  0.507487   
ABBV  0.292497  0.303988  0.419361         ...           0.291590  0.323336   
ACN   0.303224  0.395808  0.401344         ...           0.531869  0.510195   
ATVI  0.223472  0.198081  0.285139         ...           0.369710  0.398587   

            XL       XYL      YHOO       YUM       ZBH      ZION       ZTS  \
MMM   0.543725  0.623740  0.316707  0.356203  0.433634  0.486263  0.293895   
ABT   0.619808  0.567308  0.376218  0.388071  0.640906  0.512479  0.463926   
ABBV  0.267432  0.321926  0.307018  0.289727  0.508626  0.323670  0.406245   
ACN   0.563160  0.615344  0.431342  0.353530  0.461179  0.546650  0.414054   
ATVI  0.388844  0.396983  0.365830  0.330461  0.370655  0.346774  0.323145   

                     Name  
MMM            3M Company  
ABT   Abbott Laboratories  
ABBV               AbbVie  
ACN         Accenture plc  
ATVI  Activision Blizzard  

[5 rows x 497 columns]

In [34]:
def callCor2(a, b):
    cor = cor_price_name[a][b]
    name_a = cor_price_name['Name'][a]
    name_b = cor_price_name['Name'][b]
    print('The correlation between %s and %s is equal to %s' % (name_a, name_b, cor))
    
callCor2('ABBV', 'ACN')

The correlation between AbbVie and Accenture plc is equal to 0.318047748008


We ran both correlation calling functions using `%timeit` and the result is the `callCor2` with the merged table was faster with fastest time of 41.1 µs per loop, while `callCor` took approximately 1.38 ms, which is way slower.

In this part we explored the highest and lowest correlation of each company. Using `describe()` function we get the summary of each column. Using those informations, a new table containing the company ticker and their lowest and highest correlation with their corresponding tickers.

In [35]:
cor_price.describe()

MMM         ABT        ABBV         ACN        ATVI         AYI  \
count  496.000000  496.000000  496.000000  496.000000  496.000000  496.000000   
mean     0.461465    0.502471    0.318588    0.485292    0.342000    0.358049   
std      0.120494    0.114988    0.085071    0.118383    0.095141    0.086053   
min      0.161831    0.165613    0.075441    0.158036    0.003083    0.110716   
25%      0.383004    0.421799    0.267029    0.396349    0.290686    0.300497   
50%      0.466096    0.512526    0.314621    0.486101    0.347423    0.362478   
75%      0.546741    0.584136    0.363310    0.577805    0.407418    0.414250   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

             ADBE         AAP         AES         AET     ...             XEL  \
count  496.000000  496.000000  496.000000  496.000000     ...      496.000000   
mean     0.407217    0.286596    0.338180    0.337362     ...        0.337608   
std      0.108851    0.082113    0.085109    0.105296     ...        0.163069   
min      0.031734   -0.031248    0.083691    0.017283     ...        0.029882   
25%      0.340668    0.236980    0.285399    0.280438     ...        0.236055   
50%      0.414762    0.288885    0.342235    0.336069     ...        0.305559   
75%      0.486107    0.344370    0.397316    0.395577     ...        0.399817   
max      1.000000    1.000000    1.000000    1.000000     ...        1.000000   

              XRX        XLNX          XL         XYL        YHOO         YUM  \
count  496.000000  496.000000  496.000000  496.000000  496.000000  496.000000   
mean     0.423734    0.404635    0.449822    0.470841    0.306327    0.299765   
std      0.098588    0.094782    0.117506    0.120059    0.088284    0.086804   
min      0.145285    0.152021    0.121672    0.161702    0.045175    0.018138   
25%      0.350543    0.342716    0.365252    0.381844    0.260932    0.252634   
50%      0.430159    0.405600    0.455105    0.474097    0.308786    0.310659   
75%      0.497980    0.469120    0.532872    0.556206    0.365374    0.356226   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

              ZBH        ZION         ZTS  
count  496.000000  496.000000  496.000000  
mean     0.398030    0.417373    0.306473  
std      0.097566    0.159499    0.093307  
min      0.148654    0.030176    0.039721  
25%      0.335236    0.321455    0.242796  
50%      0.396175    0.415606    0.306866  
75%      0.466059    0.500168    0.363110  
max      1.000000    1.000000    1.000000  

[8 rows x 496 columns]

In [37]:
#find the bottom correlated companies
cor_bottom = cor_price.abs().idxmin(axis=0)
cor_bottom = pd.DataFrame(cor_bottom, columns = ["cor_bottom"])
#find the bottom correlation coefficients
cor_coe = cor_price.abs().min(axis=0)
cor_coe = pd.DataFrame(cor_coe, columns = ["cor_bot_coe_abs"])
#join the bottom correlation company and coefficients together
result = cor_bottom.join(cor_coe)
result.head()

cor_bottom  cor_bot_coe_abs
MMM         NEM         0.161831
ABT         NRG         0.165613
ABBV        NEM         0.075441
ACN         NEM         0.158036
ATVI        PWR         0.003083

In [38]:
#find the top correlated company
#first replace the correlation 1 to 0 to find the top correlated company
cor_modify = cor_price.replace(1,-1)
cor_top = cor_modify.idxmax(axis=0) 
cor_top = pd.DataFrame(cor_top, columns = ["cor_top"])
top_coe = cor_modify.max(axis=0)
top_coe = pd.DataFrame(top_coe, columns = ["top_coe"])
result2 = cor_top.join(top_coe)
result2.head()

cor_top   top_coe
MMM      ITW  0.768481
ABT      MMC  0.737158
ABBV     TMO  0.575175
ACN      MMC  0.725520
ATVI      EA  0.561049

In [39]:
# Then we combine the top and bottom company together 
related = result.join(result2)
related.head()

cor_bottom  cor_bot_coe_abs cor_top   top_coe
MMM         NEM         0.161831     ITW  0.768481
ABT         NRG         0.165613     MMC  0.737158
ABBV        NEM         0.075441     TMO  0.575175
ACN         NEM         0.158036     MMC  0.725520
ATVI        PWR         0.003083      EA  0.561049

The maximum and minimum correlation table was created. Thus, we could find the correlation of five tech companies, i.e. Amazon, Microsoft, Facebook, Apple and Google. The tickers of these companies are AMZN, MSFT, FB, AAPL, and GOOG/GOOGL (Google has two for Class A and Class C shares) respectively.

In [40]:
techCor = related.ix[['AMZN', 'MSFT', 'FB', 'AAPL', 'GOOG', 'GOOGL'],:]
techCor

cor_bottom  cor_bot_coe_abs cor_top   top_coe
AMZN        SRCL         0.056440   GOOGL  0.585572
MSFT        SRCL         0.028884     MMC  0.604546
FB           NEM         0.002831    FISV  0.619677
AAPL         RRC         0.112708     ITW  0.601265
GOOG        SRCL         0.017145   GOOGL  0.989365
GOOGL        RIG         0.009515    GOOG  0.989365

In [41]:
for j in list(techCor.index):
    callCor2(j, techCor.ix[j,0])
    callCor2(j, techCor.ix[j,2])

The correlation between Amazon.com Inc and Stericycle Inc is equal to 0.0564404668271
The correlation between Amazon.com Inc and Alphabet Inc Class A is equal to 0.585572457937
The correlation between Microsoft Corp. and Stericycle Inc is equal to 0.0288841252305
The correlation between Microsoft Corp. and Marsh & McLennan is equal to 0.604546183629
The correlation between Facebook and Newmont Mining Corp. (Hldg. Co.) is equal to -0.00283066410658
The correlation between Facebook and Fiserv Inc is equal to 0.619677467282
The correlation between Apple Inc. and Range Resources Corp. is equal to 0.112708151299
The correlation between Apple Inc. and Illinois Tool Works is equal to 0.601265458624
The correlation between Alphabet Inc Class C and Stericycle Inc is equal to 0.0171445946964
The correlation between Alphabet Inc Class C and Alphabet Inc Class A is equal to 0.989365335424
The correlation between Alphabet Inc Class A and Transocean is equal to 0.0095152561825
The correlation betwee

We obtained the maximum and minimum correlation of the 5 technology companies. Google who trades with two tickers (GOOG and GOOGL) will always be highly correlated to each other since the two stocks are essentially identical despite the difference in voting rights. However, their lowest correlation are different although both companies comes from a different industry. Two tech giants, Amazon Inc and Google stocks have relatively high correlation to each other with a correlation of 0.59, which is not too surprising since they released almost the same products recently.

The biggest surprise might come from Apple that was closely related to Illinois Tool Works (ITW), a manufacturing company. However, ITW also produces small electronic components, which made the correlation less surprising. Moreover, Facebook's maximum correlation came from a company called Fiserv, a financial technology company. Online payment might be an area that Facebook looking to explore, which could make sense of the correlation. Interestingly, the majority of the Tech companies had their lowest correlation against companies operating in natural resources, especially Oil and Gas.


# 3. Clustering algorithm

In [ ]:
cor_price_graph = cor_price.copy()
cor_price_graph.head()

In [ ]:
for i in range(0,cor_price_graph.shape[0]):
    for j in range(i,cor_price_graph.shape[1]):
            cor_price_graph.iloc[j,i] = 0
cor_price_graph.head()

In [ ]:
#save the graph to csv file
cor_price_graph.to_csv('cor_price_graph.csv')

** 1.Represent the edge**

In [ ]:
edge = [] # make the edge a list so that we can sort it more conveniently
for i in cor_price_graph.columns[0:cor_price_graph.shape[0]]:
    for j in cor_price_graph.index[0:cor_price_graph.shape[1]]:
        if(cor_price_graph[i][j] == 0):
            pass
        else:
            t = (i,j,cor_price_graph[i][j])  # denote the edge with (source, destination, weight)
            edge.append(t)

In [ ]:
len(edge)

so there are 122760 edges in total

In [ ]:
# sorted the edge by weight(correlation)
edge_sorted = sorted(edge, key = lambda edge: edge[2], reverse=True)
edge_sorted

** 2. Model the sets of nodes**

First, initialize each node points to itself

In [ ]:
nodePointers = {}
length = cor_price_graph.shape[0]
for i in cor_price_graph.columns[0:length]:
    nodePointers[i] = i
nodePointers

In [ ]:
set_dict = {}
for i in cor_price_graph.columns[0:3]:
    set_dict[i] = i
set_dict

In [ ]:
for i in cor_price_graph.columns[0:3]:
    print(type(i))